src/ppp/pandas.py

```diff
@@ -1,9 +1,114 @@
+import pandas as pd
 from pandas import DataFrame
 
+from ppp.common import PaymentType, ROUTE_COLUMNS
 
-def add_features(df: DataFrame) -> DataFrame:
+
+def add_features(trip_df: DataFrame, zone_df: DataFrame) -> tuple[DataFrame, DataFrame]:
+    """Returns a pandas DataFrame containing
+    all the features derived from NYC
+    Taxi dataset to answer all business questions.
+    """
+    trip_df = rename_columns_as_lowercase(trip_df)
+    zone_df = rename_columns_as_lowercase(zone_df)
+    trip_df = update_payment_type_as_string_values(trip_df)
+    trip_df = add_borough_and_zone(trip_df, zone_df, "pulocationid")
+    trip_df = add_borough_and_zone(trip_df, zone_df, "dolocationid")
+    return trip_df, zone_df
+
+
+def rename_columns_as_lowercase(df: DataFrame) -> DataFrame:
+    """Returns a pandas DataFrame containing
+    the df DataFrame with all columns
+    renamed to be lowercase.
+    """
+
+    return df.rename(columns={col: col.lower() for col in df.columns})
+
+
+def update_payment_type_as_string_values(trip_df: DataFrame) -> DataFrame:
     """Returns a pandas DataFrame containing
-    all the features required for the NYC
-    Taxi dataset.
+    the trip_df DataFrame with the payment_type
+    column updated to be string values.
     """
-    ...
+    # TODO: replace is slow, could be replaced with map
+
+    map_dict = {i.value: i.name for i in PaymentType}
```

Comment:
I like this dict comp more than my original implementation (I didn't know enums were iterable), maybe you could implement this in the polars module too?

Reply:
done